packages/webrtc-core/src/media/track.ts

```diff
@@ -0,0 +1,199 @@
+/* eslint-disable no-underscore-dangle */
+import { EventMap } from 'typed-emitter';
+import { EventEmitter } from '../event-emitter';
+import { logger } from '../util/logger';
+
+export enum Events {
+  Ended = 'ended',
+  Muted = 'muted',
+  /**
+   * Fires when the published state of a LocalTrack changes.  A value of 'true' indicates the
+   * track has been published, and 'false' indicates that it is no longer published.  Tracks are
+   * unpublished by default, and must be published explicitly (no event will be fired to indicate
+   * the initial state of 'unpublished').
+   */
+  PublishedStateUpdate = 'published-state-update',
+  /**
+   * Fires when there has been a change in the underlying track.
+   */
+  UnderlyingTrackChange = 'underlying-track-change',
+}
+
+export interface TrackState {
+  id: string;
+  label: string;
+  muted: boolean;
+}
+
+export type TrackEndEvent = {
+  trackState: TrackState;
+};
+
+export type TrackMuteEvent = {
+  trackState: TrackState;
+};
+// eslint-disable-next-line no-shadow
+export enum TrackStatus {
+  ENDED = 'ended',
+  LIVE = 'live',
+}
+
+// eslint-disable-next-line no-shadow
+export enum TrackKind {
+  AUDIO = 'audio',
+  VIDEO = 'video',
+}
+
+export type TrackPublishEvent = {
+  isPublished: boolean;
+  trackState: TrackState;
+};
+
+export interface TrackEvents extends EventMap {
+  [Events.Ended]: (event: TrackEndEvent) => void;
+  [Events.Muted]: (event: TrackMuteEvent) => void;
+  [Events.UnderlyingTrackChange]: () => void;
+}
+
+export type TrackEffect = any;
+// TBD: Fix this once types are published seperatly
+// export type TrackEffect = BaseMicrophoneEffect | BaseCameraEffect;
+
+/**
+ * Basic Track class.
+ */
+export abstract class Track extends EventEmitter<TrackEvents> {
+  static Events = Events;
+
+  ID: string;
+
+  kind: string;
+
+  status: TrackStatus;
+
+  label: string;
+
+  #mediaStreamTrack: MediaStreamTrack;
+
+  #mediaStreamTrackWithEffect?: MediaStreamTrack;
+
+  /**
+   * Constructor for the Track class. Creates an empty CoreLocalTrack or uses an existing one.
+   *
+   * @param stream - The MediaStream for this LocalTrack.
+   * @param track
+   */
+  constructor(track: MediaStreamTrack) {
+    super();
+    this.ID = track.id;
+    this.status = track.readyState as TrackStatus;
+    this.label = track.label;
+    this.kind = track.kind;
+    this.#mediaStreamTrack = track;
```

Comment:
Should we also assign `mediaStreamTrackWithEffect` to `track` here and make it not optional? In the original code, `LocalTrack` assigns both `originalStream` and `underlyingStream` in the constructor.

Reply:
we do have  #mediaStreamTrackWithEffect?: MediaStreamTrack; in the track but not optional as a developer might not need effects in some cases so that would be null

Reply:
@bbaldino  do you have a better name for this ?

Reply:
cc @marcin-bazyl

Reply:
I agree with Bryce it would be nice to be able to just call a single method and always get a track (with effect if effects are applied, without it if there are no effects). I think we can keep existing name `getMediaStreamTrackWithEffects()` and just return the `track` if there are no effects.

Reply:
ok

Reply:
@brycetham  will change it